### PR TITLE
Cloud event should parse smaller ms precisions

### PR DESCRIPTION
**Pull request by rakshith91**
Fixes https://github.com/Azure/azure-sdk-for-python/issues/19248

UPDATE:

Added more tests - this is an expected behavior

### CODE REVIEW

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 1]
```diff
@@ -146,6 +146,108 @@ def test_cloud_custom_dict_with_extensions():
...
+    assert event.data == {"team": "event grid squad"}
+    assert event.__class__ == CloudEvent
+    assert event.time.month == 2
+    assert event.time.day == 18
+    assert event.time.hour == 20
+    assert event.time.microsecond == 123
+
+def test_cloud_custom_dict_ms_precision_is_eq_six_z_not():
+    cloud_custom_dict_with_extensions = {
+        "id":"de0fd76c-4ef4-4dfb-ab3a-8f24a307e033",
```
Comment (rakshith91) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:236]:
```suggestion
        "id":"de0fd76c-4ef4-4dfb-ab3a-8f24a307e034",
```

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 2]
```diff
@@ -146,6 +146,108 @@ def test_cloud_custom_dict_with_extensions():
...
+        "time":"2021-02-18T20:18:10.123+00:00",
+        "specversion":"1.0",
+    }
+    event = CloudEvent.from_dict(cloud_custom_dict_with_extensions)
+    assert event.data == {"team": "event grid squad"}
+    assert event.__class__ == CloudEvent
+    assert event.time.month == 2
+    assert event.time.day == 18
+    assert event.time.hour == 20
+    assert event.time.microsecond == 123
```
Comment (lmazuel) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:181]: That seems unexpected compared to the regular parser in Python 3
```python
In [3]: t = datetime.fromisoformat("2021-02-18T20:18:10.123+00:00")

In [4]: t.microsecond
Out[4]: 123000
```

**Review by lmazuel — COMMENTED** [review]
I'm confused a bit, I need to re-read RFC3339 or ISO8601

File: sdk/core/azure-core/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 ## 1.15.1 (Unreleased)
 
+### Bug Fixes
```
Comment (rakshith91) [inline sdk/core/azure-core/CHANGELOG.md:5]:
```suggestion
```

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 4]
```diff
@@ -146,6 +144,108 @@ def test_cloud_custom_dict_with_extensions():
...
+        "time":"2021-02-18T20:18:10.539861122+00:00",
+        "specversion":"1.0",
+    }
+    event = CloudEvent.from_dict(cloud_custom_dict_with_extensions)
+    assert event.data == {"team": "event grid squad"}
+    assert event.__class__ == CloudEvent
+    assert event.time.month == 2
+    assert event.time.day == 18
+    assert event.time.hour == 20
+    assert event.time.microsecond == 539861
```
Comment (swathipil) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:162]: qq: so the test is checking that precision is greater than 6 decimal places, right? In that case, wouldn't you want to check that the ms `==539861122`? or am I getting mixed up?

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 5]
```diff
@@ -146,6 +144,108 @@ def test_cloud_custom_dict_with_extensions():
...
+        "time":"2021-02-18T20:18:10.123456Z",
+        "specversion":"1.0",
+    }
+    event = CloudEvent.from_dict(cloud_custom_dict_with_extensions)
+    assert event.data == {"team": "event grid squad"}
+    assert event.__class__ == CloudEvent
+    assert event.time.month == 2
+    assert event.time.day == 18
+    assert event.time.hour == 20
+    assert event.time.microsecond == 123456
```
Comment (swathipil) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:247]: so is it that, no matter what, we can only check 6 digits? and the precision part comes when we're sending the event?

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py (hunk 4 above)
Comment (rakshith91) [reply sdk/core/azure-core/tests/test_messaging_cloud_event.py:162]: Is testing that when we have more than 6 decimal places, it is trimmed to 6 - python date time objects ms precision has a limit of 6 digits

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py (hunk 5 above)
Comment (rakshith91) [reply sdk/core/azure-core/tests/test_messaging_cloud_event.py:247]: Yes - no matter what, the decimal will have 6 digits

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 6]
```diff
@@ -282,3 +382,57 @@ def test_cloud_from_dict_with_invalid_extensions():
...
+def test_cloud_custom_dict_ms_precision_is_lt_six_z_not():
+    time ="2021-02-18T20:18:10.123+00:00"
+    date_obj = _convert_to_isoformat(time)
+
+    assert date_obj.month == 2
+    assert date_obj.day == 18
+    assert date_obj.hour == 20
+    assert date_obj.microsecond == 123000
+
+def test_cloud_custom_dict_ms_precision_is_eq_six_z_not():
```
Comment (swathipil) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:431]: looks like times in these don't end with the Z?

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 7]
```diff
@@ -282,3 +382,57 @@ def test_cloud_from_dict_with_invalid_extensions():
...
+    time ="2021-02-18T20:18:10.123456+00:00"
+    date_obj = _convert_to_isoformat(time)
+
+    assert date_obj.month == 2
+    assert date_obj.day == 18
+    assert date_obj.hour == 20
+    assert date_obj.microsecond == 123456
+
+def test_cloud_custom_dict_ms_precision_is_gt_six_z_not():
+    time ="2021-02-18T20:18:10.539861122+00:00"
```
Comment (rakshith91) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:414]:
```suggestion
    time ="2021-02-18T20:18:10.539861122Z"
```

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 8]
```diff
@@ -282,3 +382,57 @@ def test_cloud_from_dict_with_invalid_extensions():
...
+    date_obj = _convert_to_isoformat(time)
+
+    assert date_obj.month == 2
+    assert date_obj.day == 18
+    assert date_obj.hour == 20
+    assert date_obj.microsecond == 539861
+
+def test_cloud_custom_dict_ms_precision_is_lt_six_z_not():
+    time ="2021-02-18T20:18:10.123+00:00"
```
Comment (rakshith91) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:423]:
```suggestion
    time ="2021-02-18T20:18:10.123Z"
```

File: sdk/core/azure-core/tests/test_messaging_cloud_event.py [hunk 9]
```diff
@@ -282,3 +382,57 @@ def test_cloud_from_dict_with_invalid_extensions():
...
+    date_obj = _convert_to_isoformat(time)
+
+    assert date_obj.month == 2
+    assert date_obj.day == 18
+    assert date_obj.hour == 20
+    assert date_obj.microsecond == 123000
+
+def test_cloud_custom_dict_ms_precision_is_eq_six_z_not():
+    time ="2021-02-18T20:18:10.123456+00:00"
```
Comment (rakshith91) [inline sdk/core/azure-core/tests/test_messaging_cloud_event.py:432]:
```suggestion
    time ="2021-02-18T20:18:10.123456Z"
```

**Review by swathipil — APPROVED** [review]
lgtm

**Pull request merged**